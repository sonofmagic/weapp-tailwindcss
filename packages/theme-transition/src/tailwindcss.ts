import type { Config } from 'tailwindcss'
import type { CSSRuleObject } from 'tailwindcss/types/config'
import defu from 'defu'
import plugin from 'tailwindcss/plugin'

export interface ThemeTransitionPluginOptions {
  zIndex?: {
    ceiling?: string | number
    floor?: string | number
  }
  // darkSelector?: string
  baseCss?: CSSRuleObject
  darkMode?: Config['darkMode']
}

type DeepRequired<T> = Required<{
  [K in keyof T]: T[K] extends Required<T[K]> ? T[K] : DeepRequired<T[K]>
}>

// https://github.com/tailwindlabs/tailwindcss/blob/557ed8ccecccf703296c186b6e996edf9933e1b7/packages/tailwindcss/src/compat/plugin-api.ts#L28
// https://github.com/tailwindlabs/tailwindcss/blob/557ed8ccecccf703296c186b6e996edf9933e1b7/packages/tailwindcss/src/compat/plugin-api.ts#L92
// https://github.com/tailwindlabs/tailwindcss/tree/main
// https://github.com/tailwindlabs/tailwindcss/blob/557ed8ccecccf703296c186b6e996edf9933e1b7/packages/tailwindcss/src/compat/dark-mode.ts#L4
export function themeTransitionPlugin(themeTransitionPluginOptions?: ThemeTransitionPluginOptions) {
  const { baseCss, zIndex, darkMode: specDarkMode } = defu<
    DeepRequired<ThemeTransitionPluginOptions>,
    ThemeTransitionPluginOptions[]
  >(
    themeTransitionPluginOptions,
    {
      baseCss: {
        mixBlendMode: 'normal',
        animation: 'none',
      },
      zIndex: {
        ceiling: 2147483646,
        floor: 1,
      },
    },
  )
  const floor = zIndex.floor.toString()
  const ceiling = zIndex.ceiling.toString()
  return plugin(({ addBase, config }) => {
    const darkMode: Config['darkMode'] = specDarkMode ?? config('darkMode')
    let [mode, selector = '.dark'] = Array.isArray(darkMode) ? darkMode : [darkMode]
    if (mode === 'variant') {
      let formats

      if (Array.isArray(selector)) {
        formats = selector
      }
      else if (typeof selector === 'function') {
        formats = selector
      }
      else if (typeof selector === 'string') {
        formats = [selector]
      }

      if (Array.isArray(formats)) {
        for (const format of formats) {
          if (format === '.dark') {
            mode = undefined
            console.warn(
              'When using `variant` for `darkMode`, you must provide a selector.\nExample: `darkMode: ["variant", ".your-selector &"]`',
            )
          }
          else if (!format.includes('&')) {
            mode = undefined
            console.warn(
              'When using `variant` for `darkMode`, your selector must contain `&`.\nExample `darkMode: ["variant", ".your-selector &"]`',
            )
          }
        }
      }

      selector = formats as any
    }
    const rule: CSSRuleObject = {
      '::view-transition-old(root),\n::view-transition-new(root)': baseCss,
      '::view-transition-old(root)': {
        zIndex: floor,
      },
      '::view-transition-new(root)': {
        zIndex: ceiling,
      },
    }
    if (mode === 'selector' || mode === 'class') {
      rule[`${selector}::view-transition-old(root)`] = {
        zIndex: ceiling,
      }
      rule[`${selector}::view-transition-new(root)`] = {
        zIndex: floor,
      }
    }
    else if (mode === 'media') {
      rule['@media (prefers-color-scheme: dark)'] = {
        '::view-transition-old(root)': {
          zIndex: ceiling,
        },
        '::view-transition-new(root)': {
          zIndex: floor,
        },
      }
    }
    // || mode === 'variant'

    addBase(
      rule,
    )
  })
}

export default themeTransitionPlugin
