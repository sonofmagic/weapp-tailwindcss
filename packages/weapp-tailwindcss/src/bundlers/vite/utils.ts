import type { ExistingRawSourceMap } from 'rollup'
import path from 'node:path'
import process from 'node:process'

export function slash(p: string): string {
  return p.replace(/\\/g, '/')
}

export const isWindows = process.platform === 'win32'

const cssLangs = `\\.(css|less|sass|scss|styl|stylus|pcss|postcss)($|\\?)`
export const cssLangRE = new RegExp(cssLangs)
export function isCSSRequest(request: string): boolean {
  return cssLangRE.test(request)
}

export function normalizePath(id: string): string {
  return path.posix.normalize(isWindows ? slash(id) : id)
}

const postfixRE = /[?#].*$/
export function cleanUrl(url: string): string {
  return url.replace(postfixRE, '')
}

export async function formatPostcssSourceMap(
  rawMap: ExistingRawSourceMap,
  file: string,
): Promise<ExistingRawSourceMap> {
  const inputFileDir = path.dirname(file)

  const sources = rawMap.sources.map((source) => {
    const cleanSource = cleanUrl(decodeURIComponent(source))

    // postcss virtual files
    if (cleanSource[0] === '<' && cleanSource.endsWith('>')) {
      return `\0${cleanSource}`
    }

    return normalizePath(path.resolve(inputFileDir, cleanSource))
  })

  return {
    file,
    mappings: rawMap.mappings,
    names: rawMap.names,
    sources,
    sourcesContent: rawMap.sourcesContent,
    version: rawMap.version,
  }
}
