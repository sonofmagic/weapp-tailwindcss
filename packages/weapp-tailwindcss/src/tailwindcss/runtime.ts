import type { TailwindcssPatcherLike } from '@/types'
import { createDebug } from '@/debug'

const debug = createDebug('[tailwindcss:runtime] ')

function shouldPreferSync(majorVersion: number | undefined) {
  if (majorVersion == null) {
    return true
  }
  if (majorVersion === 3) {
    return true
  }
  if (majorVersion === 4) {
    return true
  }
  return false
}

function tryGetRuntimeClassSetSync(twPatcher: TailwindcssPatcherLike) {
  if (typeof twPatcher.getClassSetSync !== 'function') {
    return undefined
  }

  if (!shouldPreferSync(twPatcher.majorVersion)) {
    return undefined
  }

  try {
    return twPatcher.getClassSetSync()
  }
  catch (error) {
    if (twPatcher.majorVersion === 4) {
      debug('getClassSetSync() unavailable for tailwindcss v4, fallback to async getClassSet(): %O', error)
    }
    else {
      debug('getClassSetSync() failed, fallback to async getClassSet(): %O', error)
    }
    return undefined
  }
}

async function collectRuntimeClassSet(twPatcher: TailwindcssPatcherLike): Promise<Set<string>> {
  const syncSet = tryGetRuntimeClassSetSync(twPatcher)
  if (syncSet) {
    return syncSet
  }

  try {
    const result = await twPatcher.extract({ write: false })
    if (result?.classSet) {
      return result.classSet
    }
  }
  catch (error) {
    debug('extract() failed, fallback to getClassSet(): %O', error)
  }

  try {
    const fallbackSet = await Promise.resolve(twPatcher.getClassSet())
    if (fallbackSet) {
      return fallbackSet
    }
  }
  catch (error) {
    debug('getClassSet() failed, returning empty set: %O', error)
  }

  return new Set<string>()
}

export { collectRuntimeClassSet }
