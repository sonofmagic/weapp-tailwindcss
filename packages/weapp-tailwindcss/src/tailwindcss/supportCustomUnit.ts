import * as t from '@babel/types'
import type { ArrayExpression, StringLiteral } from '@babel/types'
import { parse, traverse } from '@/babel'
import type { ILengthUnitsPatchOptions, ILengthUnitsPatchDangerousOptions } from '@/types'

export function findAstNode(content: string, options: ILengthUnitsPatchOptions) {
  const DOPTS = options.dangerousOptions as Required<ILengthUnitsPatchDangerousOptions>
  const ast = parse(content)

  let arrayRef: ArrayExpression | undefined
  let changed = false
  traverse(ast, {
    Identifier(path) {
      if (
        path.node.name === DOPTS.variableName &&
        t.isVariableDeclarator(path.parent) &&
        t.isArrayExpression(path.parent.init)
      ) {
        arrayRef = path.parent.init
        const set = new Set(path.parent.init.elements.map((x) => (<StringLiteral>x).value))
        for (let i = 0; i < options.units.length; i++) {
          const unit = options.units[i]
          if (!set.has(unit)) {
            path.parent.init.elements = path.parent.init.elements.map((x) => {
              if (t.isStringLiteral(x)) {
                return {
                  type: x?.type,
                  value: x?.value
                }
              }
              return x
            })
            path.parent.init.elements.push({
              type: 'StringLiteral',
              value: unit
            })
            changed = true
          }
        }
      }
    }
  })
  return {
    arrayRef,
    changed
  }
}
