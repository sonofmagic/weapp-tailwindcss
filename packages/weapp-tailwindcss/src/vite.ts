export * from './bundlers/vite/index'
