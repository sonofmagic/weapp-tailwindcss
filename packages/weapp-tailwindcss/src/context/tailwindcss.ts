import type { TailwindcssPatchOptions } from 'tailwindcss-patch'
import type { InternalUserDefinedOptions } from '@/types'
import { createRequire } from 'node:module'
import path from 'node:path'
import process from 'node:process'
import { fileURLToPath } from 'node:url'
import { logger } from '@weapp-tailwindcss/logger'
import { createTailwindcssPatcher } from '@/tailwindcss'
import { defuOverrideArray } from '@/utils'

const ENV_BASEDIR_KEYS = [
  'WEAPP_TAILWINDCSS_BASEDIR',
  'WEAPP_TAILWINDCSS_BASE_DIR',
  'TAILWINDCSS_BASEDIR',
  'TAILWINDCSS_BASE_DIR',
  'UNI_INPUT_DIR',
  'UNI_INPUT_ROOT',
  'UNI_CLI_ROOT',
  'UNI_APP_INPUT_DIR',
  'INIT_CWD',
  'PWD',
] as const

type EnvBasedirKey = typeof ENV_BASEDIR_KEYS[number]

const GENERIC_ENV_BASEDIR_KEYS = new Set<EnvBasedirKey>(['INIT_CWD', 'PWD'])

interface EnvBasedirResult {
  key: EnvBasedirKey
  value: string
}

function pickEnvBasedir(): EnvBasedirResult | undefined {
  for (const key of ENV_BASEDIR_KEYS) {
    const value = process.env[key]
    if (value && path.isAbsolute(value)) {
      return { key, value }
    }
  }
  return undefined
}

function pickPackageEnvBasedir(): string | undefined {
  const packageJsonPath = process.env.npm_package_json
  if (packageJsonPath) {
    const packageDir = path.dirname(packageJsonPath)
    if (packageDir && path.isAbsolute(packageDir)) {
      return packageDir
    }
  }
  const localPrefix = process.env.npm_config_local_prefix
  if (localPrefix && path.isAbsolute(localPrefix)) {
    return localPrefix
  }
  return undefined
}

function detectCallerBasedir(): string | undefined {
  const stack = new Error('resolveTailwindcssBasedir stack probe').stack
  if (!stack) {
    return undefined
  }
  if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
    logger.debug('caller stack: %s', stack)
  }
  const lines = stack.split('\n')
  for (const line of lines) {
    const match = line.match(/\(([^)]+)\)/u) ?? line.match(/at\s+(\S.*)$/u)
    const location = match?.[1]
    if (!location) {
      continue
    }
    let filePath = location
    if (filePath.startsWith('file://')) {
      try {
        filePath = fileURLToPath(filePath)
      }
      catch {
        continue
      }
    }
    const [candidate] = filePath.split(':')
    const resolvedPath = path.isAbsolute(filePath) ? filePath : candidate
    if (!path.isAbsolute(resolvedPath)) {
      continue
    }
    if (resolvedPath.includes('node_modules') && resolvedPath.includes('weapp-tailwindcss')) {
      continue
    }
    try {
      return path.dirname(resolvedPath)
    }
    catch {
      continue
    }
  }
  return undefined
}

export function resolveTailwindcssBasedir(basedir?: string) {
  const envBasedirResult = pickEnvBasedir()
  const envBasedir = envBasedirResult?.value
  const envBasedirKey = envBasedirResult?.key
  const envBasedirIsGeneric = envBasedirKey ? GENERIC_ENV_BASEDIR_KEYS.has(envBasedirKey) : false
  const packageEnvBasedir = pickPackageEnvBasedir()
  const shouldDetectCaller = !envBasedir || envBasedirIsGeneric
  const callerBasedir = shouldDetectCaller ? detectCallerBasedir() : undefined
  const anchor = envBasedir ?? packageEnvBasedir ?? callerBasedir ?? process.cwd()
  if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
    logger.debug('resolveTailwindcssBasedir anchor %O', {
      basedir,
      envBasedir,
      envBasedirKey,
      envBasedirIsGeneric,
      packageEnvBasedir,
      callerBasedir,
      npm_package_json: process.env.npm_package_json,
      cwd: process.cwd(),
      anchor,
    })
  }

  if (basedir && basedir.trim().length > 0) {
    if (path.isAbsolute(basedir)) {
      return path.normalize(basedir)
    }
    return path.resolve(anchor, basedir)
  }

  if (envBasedir && !envBasedirIsGeneric) {
    return path.normalize(envBasedir)
  }

  if (packageEnvBasedir) {
    return path.normalize(packageEnvBasedir)
  }

  if (callerBasedir) {
    const normalizedCaller = path.normalize(callerBasedir)
    const librarySegment = `${path.sep}weapp-tailwindcss${path.sep}`
    if (!normalizedCaller.includes(librarySegment)) {
      return normalizedCaller
    }
  }

  if (envBasedir) {
    return path.normalize(envBasedir)
  }

  const packageName = process.env.PNPM_PACKAGE_NAME
  if (packageName) {
    try {
      const anchorRequire = createRequire(path.join(anchor, '__resolve_tailwindcss_basedir__.cjs'))
      const packageJsonPath = anchorRequire.resolve(`${packageName}/package.json`)
      if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
        logger.debug('package basedir resolved from PNPM_PACKAGE_NAME: %s', packageJsonPath)
      }
      return path.normalize(path.dirname(packageJsonPath))
    }
    catch {
      if (process.env.WEAPP_TW_DEBUG_STACK === '1') {
        logger.debug('failed to resolve package json for %s', packageName)
      }
      // ignore
    }
  }

  return path.normalize(process.cwd())
}

export function createTailwindcssPatcherFromContext(ctx: InternalUserDefinedOptions) {
  const {
    tailwindcssBasedir,
    supportCustomLengthUnitsPatch,
    tailwindcss,
    tailwindcssPatcherOptions,
    cssEntries,
    appType,
  } = ctx

  const resolvedTailwindcssBasedir = resolveTailwindcssBasedir(tailwindcssBasedir)
  ctx.tailwindcssBasedir = resolvedTailwindcssBasedir
  logger.debug('tailwindcss basedir resolved: %s', resolvedTailwindcssBasedir)

  type TailwindUserOptions = NonNullable<TailwindcssPatchOptions['tailwind']>

  const defaultTailwindcssConfig: TailwindUserOptions = {
    cwd: resolvedTailwindcssBasedir,
    v2: {
      cwd: resolvedTailwindcssBasedir,
    },
    v3: {
      cwd: resolvedTailwindcssBasedir,
    },
    v4: {
      base: resolvedTailwindcssBasedir,
      cssEntries,
    },
  }

  if (cssEntries?.length && (tailwindcss == null || tailwindcss.version == null)) {
    defaultTailwindcssConfig.version = 4
  }

  return createTailwindcssPatcher(
    {
      basedir: resolvedTailwindcssBasedir,
      cacheDir: appType === 'mpx' ? 'node_modules/tailwindcss-patch/.cache' : undefined,
      supportCustomLengthUnitsPatch: supportCustomLengthUnitsPatch ?? true,
      tailwindcss: defuOverrideArray<TailwindUserOptions, TailwindUserOptions[]>(
        (tailwindcss ?? {}) as TailwindUserOptions,
        defaultTailwindcssConfig,
      ),
      tailwindcssPatcherOptions,
    },
  )
}
