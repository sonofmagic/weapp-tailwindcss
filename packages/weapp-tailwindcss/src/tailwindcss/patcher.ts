import path from 'node:path'
import fs from 'node:fs'
import process from 'node:process'
import { gte as semverGte } from 'semver'
import type { PackageJson } from 'pkg-types'
import { TailwindcssPatcher, monkeyPatchForExposingContext, requireResolve } from 'tailwindcss-patch'
import { findAstNode } from './supportCustomUnit'
import type { ILengthUnitsPatchDangerousOptions, ILengthUnitsPatchOptions, InternalPatchResult } from '@/types'
import { noop } from '@/utils'
import { pluginName } from '@/constants'
import { generate } from '@/babel'

// `${cwd}/node_modules/${dangerousOptions.packageName}/package.json`
// const pkgJson = require(tmpJsonPath) as PackageJson
// https://github.com/sonofmagic/weapp-tailwindcss-webpack-plugin/issues/110
// https://github.com/tailwindlabs/tailwindcss/discussions/9675
// if (semverGte(pkgJson.version!, dangerousOptions.gteVersion)) {
// }
// const cwd = process.cwd()
export function getInstalledPkgJsonPath(options: ILengthUnitsPatchOptions) {
  const dangerousOptions = options.dangerousOptions as Required<ILengthUnitsPatchDangerousOptions>
  try {
    const tmpJsonPath = requireResolve(`${dangerousOptions.packageName}/package.json`, {
      paths: options.paths,
      basedir: options.basedir,
    })

    return tmpJsonPath
  }
  catch (error) {
    if ((<Error & { code: string }>error).code === 'MODULE_NOT_FOUND') {
      console.warn(
        '没有找到`tailwindcss`包，请确认是否安装。想要禁用打上rpx支持patch或者非`tailwindcss`框架，你可以设置 `supportCustomLengthUnitsPatch` 为 false',
      )
    }
  }
}

export function createPatch(options: false | ILengthUnitsPatchOptions) {
  if (options === false) {
    return noop
  }
  return () => {
    try {
      return internalPatch(getInstalledPkgJsonPath(options), options)
    }
    catch (error) {
      console.warn(`patch tailwindcss failed:${(<Error>error).message}`)
    }
  }
}

export function monkeyPatchForSupportingCustomUnit(rootDir: string, options: ILengthUnitsPatchOptions) {
  const { dangerousOptions } = options
  const DOPTS = dangerousOptions as Required<ILengthUnitsPatchDangerousOptions>
  const dataTypesFilePath = path.resolve(rootDir, DOPTS.lengthUnitsFilePath)
  const dataTypesFileContent = fs.readFileSync(dataTypesFilePath, {
    encoding: 'utf8',
  })
  const { arrayRef, changed } = findAstNode(dataTypesFileContent, options)
  if (arrayRef && changed) {
    const { code } = generate(arrayRef, {
      jsescOption: {
        quotes: 'single',
      },
    })

    if (arrayRef.start && arrayRef.end) {
      const prev = dataTypesFileContent.slice(0, arrayRef.start)
      const next = dataTypesFileContent.slice(arrayRef.end as number)
      const newCode = prev + code + next
      if (DOPTS.overwrite) {
        fs.writeFileSync(DOPTS.destPath ?? dataTypesFilePath, newCode, {
          encoding: 'utf8',
        })
        console.log('patch tailwindcss for custom length unit successfully!')
      }
    }
    return code
  }
}

export function internalPatch(
  pkgJsonPath: string | undefined,
  options: ILengthUnitsPatchOptions,
  overwrite = true,
): InternalPatchResult | undefined {
  if (pkgJsonPath) {
    const pkgJson = JSON.parse(fs.readFileSync(pkgJsonPath, 'utf8')) as PackageJson
    const dangerousOptions = options.dangerousOptions as Required<ILengthUnitsPatchDangerousOptions>
    const version = pkgJson.version!
    if (semverGte(version, dangerousOptions.gteVersion)) {
      const rootDir = path.dirname(pkgJsonPath)
      const dataTypes = monkeyPatchForSupportingCustomUnit(rootDir, options)
      const result = monkeyPatchForExposingContext(rootDir, {
        overwrite,
        version,
      })
      return {
        ...result,
        dataTypes,
      }
    }
  }
}

export function mkCacheDirectory(cwd = process.cwd()) {
  const cacheDirectory = path.resolve(cwd, 'node_modules', '.cache', pluginName)

  const exists = fs.existsSync(cacheDirectory)
  if (!exists) {
    fs.mkdirSync(cacheDirectory, {
      recursive: true,
    })
  }
  return cacheDirectory
}

export function createTailwindcssPatcher() {
  return new TailwindcssPatcher({
    cache: true,
  })
}
