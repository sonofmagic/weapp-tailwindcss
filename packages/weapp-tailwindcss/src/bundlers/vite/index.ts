import type { Plugin } from 'vite'
import type { OutputAsset, OutputChunk } from 'rollup'
import type { UserDefinedOptions } from '@/types'
import { getOptions } from '@/options'
import { vitePluginName } from '@/constants'
import { getGroupedEntries } from '@/utils'
import { createTailwindcssPatcher } from '@/tailwindcss/patcher'
import { createDebug } from '@/debug'

const debug = createDebug()

/**
 * @name UnifiedViteWeappTailwindcssPlugin
 * @description uni-app vite vue3 版本插件
 * @link https://weapp-tw.icebreaker.top/docs/quick-start/frameworks/uni-app-vite
 */
export function UnifiedViteWeappTailwindcssPlugin(options: UserDefinedOptions = {}): Plugin | undefined {
  const opts = getOptions(options)
  const {
    disabled,
    onEnd,
    onLoad,
    onStart,
    onUpdate,
    templateHandler,
    styleHandler,
    patch,
    jsHandler,
    mainCssChunkMatcher,
    appType,
    setMangleRuntimeSet,
    cache,
    tailwindcssBasedir,
  } = opts
  if (disabled) {
    return
  }

  patch?.()

  const twPatcher = createTailwindcssPatcher()
  onLoad()
  // 要在 vite:css 处理之前运行
  return {
    name: vitePluginName,
    enforce: 'post',
    async generateBundle(opt, bundle) {
      debug('start')
      onStart()

      const entries = Object.entries(bundle)
      const groupedEntries = getGroupedEntries(entries, opts)
      const runtimeSet = twPatcher.getClassSet({
        basedir: tailwindcssBasedir,
      })
      setMangleRuntimeSet(runtimeSet)
      debug('get runtimeSet, class count: %d', runtimeSet.size)
      if (Array.isArray(groupedEntries.html)) {
        let noCachedCount = 0
        for (let i = 0; i < groupedEntries.html.length; i++) {
          const [file, originalSource] = groupedEntries.html[i] as [string, OutputAsset]

          const oldVal = originalSource.source.toString()

          const hash = cache.computeHash(oldVal)
          cache.calcHashValueChanged(file, hash)

          await cache.process(
            file,
            () => {
              const source = cache.get<string>(file)
              if (source) {
                originalSource.source = source
                debug('html cache hit: %s', file)
              }
              else {
                return false
              }
            },
            async () => {
              originalSource.source = await templateHandler(oldVal, {
                runtimeSet,
              })
              onUpdate(file, oldVal, originalSource.source)
              debug('html handle: %s', file)
              noCachedCount++
              return {
                key: file,
                source: originalSource.source,
              }
            },
          )
        }
        debug('html handle finish, total: %d, no-cached: %d', groupedEntries.html.length, noCachedCount)
      }
      if (Array.isArray(groupedEntries.js)) {
        let noCachedCount = 0
        for (let i = 0; i < groupedEntries.js.length; i++) {
          const [file, originalSource] = groupedEntries.js[i] as [string, OutputChunk]
          const rawSource = originalSource.code

          const hash = cache.computeHash(rawSource)
          cache.calcHashValueChanged(file, hash)
          await cache.process(
            file,
            () => {
              const source = cache.get<string>(file)
              if (source) {
                originalSource.code = source
                debug('js cache hit: %s', file)
              }
              else {
                return false
              }
            },
            async () => {
              const mapFilename = `${file}.map`
              const hasMap = Boolean(bundle[mapFilename])
              const { code, map } = await jsHandler(rawSource, runtimeSet, {
                generateMap: hasMap,
              })
              originalSource.code = code
              onUpdate(file, rawSource, code)
              debug('js handle: %s', file)
              noCachedCount++
              if (hasMap && map) {
                ;(bundle[mapFilename] as OutputAsset).source = map.toString()
              }
              return {
                key: file,
                source: code,
              }
            },
          )
        }
        debug('js handle finish, total: %d, no-cached: %d', groupedEntries.js.length, noCachedCount)
      }

      if (Array.isArray(groupedEntries.css)) {
        let noCachedCount = 0
        for (let i = 0; i < groupedEntries.css.length; i++) {
          const [file, originalSource] = groupedEntries.css[i] as [string, OutputAsset]

          const rawSource = originalSource.source.toString()

          const hash = cache.computeHash(rawSource)
          cache.calcHashValueChanged(file, hash)
          await cache.process(
            file,
            () => {
              const source = cache.get<string>(file)
              if (source) {
                originalSource.source = source
                debug('css cache hit: %s', file)
              }
              else {
                return false
              }
            },
            async () => {
              const css = await styleHandler(rawSource, {
                isMainChunk: mainCssChunkMatcher(originalSource.fileName, appType),
              })
              originalSource.source = css
              onUpdate(file, rawSource, css)
              debug('css handle: %s', file)
              noCachedCount++
              return {
                key: file,
                source: css,
              }
            },
          )
        }
        debug('css handle finish, total: %d, no-cached: %d', groupedEntries.css.length, noCachedCount)
      }

      onEnd()
      debug('end')
    },
  }
}
