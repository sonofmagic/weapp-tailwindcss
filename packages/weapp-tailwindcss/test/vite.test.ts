import tailwindcss from '@tailwindcss/postcss'
import twv from '@tailwindcss/vite'
import { isCI } from 'ci-info'
import fs from 'fs-extra'
import path from 'pathe'
import { build } from 'vite'
import { UnifiedViteWeappTailwindcssPlugin } from '@/vite'
import { fixturesRootPath } from './util'

describe('vite', () => {
  it.skipIf(isCI)('v4-vite-plugin', async () => {
    await build({
      root: path.resolve(fixturesRootPath, 'v4-vite-plugin'),
      plugins: [
        twv(),
        UnifiedViteWeappTailwindcssPlugin(),
      ],
      build: {
        minify: false,
        rollupOptions: {
          output: {
            assetFileNames: '[name].[ext]',
            entryFileNames: '[name].js',
            chunkFileNames: '[name].js',
          },
        },
      },
    })
    expect(await fs.readFile(path.resolve(fixturesRootPath, 'v4-vite-plugin/dist/index.css'), 'utf-8')).toMatchSnapshot('css')
    expect(await fs.readFile(path.resolve(fixturesRootPath, 'v4-vite-plugin/dist/index.js'), 'utf-8')).toMatchSnapshot('js')
  })

  it('v4-vite-postcss', async () => {
    const root = path.resolve(fixturesRootPath, 'v4-vite-postcss')
    await build({
      root,
      plugins: [
        UnifiedViteWeappTailwindcssPlugin(),
      ],
      build: {
        minify: false,
        rollupOptions: {
          output: {
            assetFileNames: '[name].[ext]',
            entryFileNames: '[name].js',
            chunkFileNames: '[name].js',
          },
        },
      },
      css: {
        postcss: {
          plugins: [
            tailwindcss(
              {
                base: root,
              },
            ),
          ],
        },
      },
    })
    expect(await fs.readFile(path.resolve(fixturesRootPath, 'v4-vite-postcss/dist/index.css'), 'utf-8')).toMatchSnapshot('css')
    expect(await fs.readFile(path.resolve(fixturesRootPath, 'v4-vite-postcss/dist/index.js'), 'utf-8')).toMatchSnapshot('js')
  })
})
