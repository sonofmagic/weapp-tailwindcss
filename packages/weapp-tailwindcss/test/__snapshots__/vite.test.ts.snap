// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vite > v4-vite-plugin > css 1`] = `
"/*! tailwindcss v4.1.16 | MIT License | https://tailwindcss.com */
view,text,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}

.bg-_h123456_ {
  background-color: #123456;
}

.text-_h123456_ {
  color: #123456;
}
"
`;

exports[`vite > v4-vite-plugin > js 1`] = `
"(function polyfill() {
  const relList = document.createElement("link").relList;
  if (relList && relList.supports && relList.supports("modulepreload")) return;
  for (const link of document.querySelectorAll('link[rel="modulepreload"]')) processPreload(link);
  new MutationObserver((mutations) => {
    for (const mutation of mutations) {
      if (mutation.type !== "childList") continue;
      for (const node of mutation.addedNodes) if (node.tagName === "LINK" && node.rel === "modulepreload") processPreload(node);
    }
  }).observe(document, {
    childList: true,
    subtree: true
  });
  function getFetchOpts(link) {
    const fetchOpts = {};
    if (link.integrity) fetchOpts.integrity = link.integrity;
    if (link.referrerPolicy) fetchOpts.referrerPolicy = link.referrerPolicy;
    if (link.crossOrigin === "use-credentials") fetchOpts.credentials = "include";
    else if (link.crossOrigin === "anonymous") fetchOpts.credentials = "omit";
    else fetchOpts.credentials = "same-origin";
    return fetchOpts;
  }
  function processPreload(link) {
    if (link.ep) return;
    link.ep = true;
    const fetchOpts = getFetchOpts(link);
    fetch(link.href, fetchOpts);
  }
})();
console.log("bg-_h123456_ text-_h123456_ first-letter:marker:text-clip");
"
`;

exports[`vite > v4-vite-postcss > css 1`] = `
"/*! tailwindcss v4.1.16 | MIT License | https://tailwindcss.com */
view,text,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}
.bg-_h123456_ {
  background-color: #123456;
}
.text-_h123456_ {
  color: #123456;
}
.first-lettercmarkerctext-clip::-csstools-invalid-first-letter view::marker {
      text-overflow: clip;
    }
.first-lettercmarkerctext-clip::-csstools-invalid-first-letter text::marker {
      text-overflow: clip;
    }
.first-lettercmarkerctext-clip::-csstools-invalid-first-letter::marker {
      text-overflow: clip;
    }
.first-lettercmarkerctext-clip::-csstools-invalid-first-letter view::-webkit-details-marker {
      text-overflow: clip;
    }
.first-lettercmarkerctext-clip::-csstools-invalid-first-letter text::-webkit-details-marker {
      text-overflow: clip;
    }
.first-lettercmarkerctext-clip::-csstools-invalid-first-letter::-webkit-details-marker {
      text-overflow: clip;
    }
"
`;

exports[`vite > v4-vite-postcss > js 1`] = `
"(function polyfill() {
  const relList = document.createElement("link").relList;
  if (relList && relList.supports && relList.supports("modulepreload")) return;
  for (const link of document.querySelectorAll('link[rel="modulepreload"]')) processPreload(link);
  new MutationObserver((mutations) => {
    for (const mutation of mutations) {
      if (mutation.type !== "childList") continue;
      for (const node of mutation.addedNodes) if (node.tagName === "LINK" && node.rel === "modulepreload") processPreload(node);
    }
  }).observe(document, {
    childList: true,
    subtree: true
  });
  function getFetchOpts(link) {
    const fetchOpts = {};
    if (link.integrity) fetchOpts.integrity = link.integrity;
    if (link.referrerPolicy) fetchOpts.referrerPolicy = link.referrerPolicy;
    if (link.crossOrigin === "use-credentials") fetchOpts.credentials = "include";
    else if (link.crossOrigin === "anonymous") fetchOpts.credentials = "omit";
    else fetchOpts.credentials = "same-origin";
    return fetchOpts;
  }
  function processPreload(link) {
    if (link.ep) return;
    link.ep = true;
    const fetchOpts = getFetchOpts(link);
    fetch(link.href, fetchOpts);
  }
})();
console.log("bg-_h123456_ text-_h123456_");
"
`;
